### PR TITLE
Validation Layers

### PR DESCRIPTION
**Pull request by Delaunay**

### CODE REVIEW

**Comment by Delaunay** [comment]
Check report examples
```
Nan Check
=========
  * Loss increased 2 times

Planning Checks
===============
  * Wrong number of configs

```

**Comment by Delaunay** [comment]
ADD warning capture

**Comment by Delaunay** [comment]
Report Example
```
resnet152.D0
============
  No traceback info about the error
  * Loss increased 5 times
  * Device 0 used memory is below threshold  0.02 <  0.50
resnet152.D1
============
  No traceback info about the error
  * Loss increased 5 times
  * Device 1 used memory is below threshold  0.02 <  0.50
resnet152-multi.0
=================
  * torch.distributed.elastic.multiprocessing.api.SignalException: Process 374271 got signal: 15
  * Loss increased 6 times
  * Device 0 used memory is below threshold  0.02 <  0.50
  * Device 1 used memory is below threshold  0.02 <  0.50
davit_large.D0
==============
  No traceback info about the error
  * Loss increased 6 times
davit_large.D1
==============
  No traceback info about the error
  * Loss increased 6 times
davit_large-multi.0
===================
  * torch.distributed.elastic.multiprocessing.api.SignalException: Process 374932 got signal: 15
  * Loss increased 8 times
focalnet.D0
===========
  No traceback info about the error
  * Loss increased 6 times
  * Device 0 used memory is below threshold  0.01 <  0.50
focalnet.D1
===========
  No traceback info about the error
  * Loss increased 6 times
  * Device 1 used memory is below threshold  0.01 <  0.50
resnet50.D0
===========
  * Loss increased 7 times
  * Device 0 used memory is below threshold  0.04 <  0.50
resnet50.D1
===========
  * Loss increased 7 times
  * Device 1 used memory is below threshold  0.04 <  0.50
convnext_large-fp32.D0
======================
  * Loss increased 17 times
convnext_large-fp32.D1
======================
  * Loss increased 17 times
convnext_large-fp16.D0
======================
  * Loss increased 17 times
  * Device 0 used memory is below threshold  0.00 <  0.50
convnext_large-fp16.D1
======================
  * Loss increased 17 times
  * Device 1 used memory is below threshold  0.00 <  0.50
convnext_large-tf32.D1
======================
  * Loss increased 17 times
convnext_large-tf32.D0
======================
  * Loss increased 17 times
convnext_large-tf32-fp16.D0
===========================
  * Loss increased 17 times
  * Device 0 used memory is below threshold  0.00 <  0.50
convnext_large-tf32-fp16.D1
===========================
  * Loss increased 17 times
  * Device 1 used memory is below threshold  0.00 <  0.50
regnet_y_128gf.D0
=================
  * Loss increased 7 times
regnet_y_128gf.D1
=================
  * Loss increased 7 times
bert-fp32.D1
============
  * Loss increased 21 times
bert-fp32.D0
============
  * Loss increased 21 times
bert-fp16.D0
============
  * Loss increased 3 times
bert-fp16.D1
============
  * Loss increased 3 times
bert-tf32.D1
============
  * Loss increased 21 times
bert-tf32.D0
============
  * Loss increased 21 times
bert-tf32-fp16.D1
=================
  * Loss increased 3 times
bert-tf32-fp16.D0
=================
  * Loss increased 3 times
t5.D1
=====
  * Loss increased 2 times
  * Device 1 used memory is below threshold  0.06 <  0.50
t5.D0
=====
  * Loss increased 2 times
  * Device 0 used memory is below threshold  0.06 <  0.50
whisper.D1
==========
  * Loss increased 6 times
whisper.D0
==========
  * Loss increased 6 times
opt-6_7b.0
==========
  * Loss increased 50 times
  * Device 0 used memory is below threshold  0.01 <  0.50
  * Device 1 used memory is below threshold  0.01 <  0.50
stargan.D0
==========
  * Loss increased 5 times
  * Device 0 used memory is below threshold  0.01 <  0.50
stargan.D1
==========
  * Loss increased 5 times
  * Device 1 used memory is below threshold  0.01 <  0.50
dlrm.0
======
  * Loss increased 3 times
  * Device 0 used memory is below threshold  0.01 <  0.50
  * Device 1 used memory is below threshold  0.01 <  0.50
 ```

File: milabench/utils.py [hunk 1]
```diff
@@ -102,3 +108,92 @@ def make_constraints_file(pth, constraints):
...
+        for _, fun in self.funs.items():
+            try:
+                fun(*args, **kwds)
+
+            except Exception:
+                logger_name = getattr(fun, "__name__", fun)
+                print(f"Error happened in logger {logger_name}", file=sys.stderr)
+                print("=" * 80)
+                traceback.print_exc()
+                print("=" * 80)
```
Comment (breuleux) [inline milabench/utils.py:167]: The =s should be printed to stderr as well (that was also an oversight in master currently, I realize)

File: milabench/utils.py [hunk 2]
```diff
@@ -102,3 +108,92 @@ def make_constraints_file(pth, constraints):
...
+    path = module.__path__
+    name = module.__name__
+
+    layers = {}
+
+    for _, layerpath, _ in pkgutil.iter_modules(path, name + "."):
+        layername = layerpath.split(".")[-1]
+        layermodule = importlib.import_module(layerpath)
+
+        if hasattr(layermodule, "_Layer"):
```
Comment (breuleux) [inline milabench/utils.py:124]: A single leading underscore typically denotes a private variable. I would make it `Layer` or `__layer__` or `__layers__ = {"layer_name": LayerClass}` if we want to allow a module to export multiple layers.

File: milabench/validation/ensure_rate.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    def on_event(self, entry: BenchLogEntry):
+        tag = entry.tag
+
+        if tag not in self.rates:
+            self.rates[tag] = 0
```
Comment (breuleux) [inline milabench/validation/ensure_rate.py:25]: That shouldn't be necessary, given that `self.rates` is a `defaultdict(float)`.

File: milabench/validation/ensure_rate.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+        if entry.event == "data":
```
Comment (breuleux) [inline milabench/validation/ensure_rate.py:27]: I feel like we could reduce boilerplate a bit by having the `ValidationLayer` class dispatch to `getattr(self, f"on_{entry.event}")`. Just a thought.

File: milabench/validation/error.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            output.append(line)
+
+    return output
+
+
+class _Layer(ValidationLayer):
+    """Capture all error event and save them to generate a summary"""
+
+    def __init__(self, **kwargs) -> None:
+        self.errors = defaultdict(PackError)
```
Comment (breuleux) [inline milabench/validation/error.py:42]: The layers almost all have some tag-indexed defaultdict, I feel like there's something about this pattern that could be abstracted, maybe have a method to initialize for a new tag, and then have on_event take that value for the current tag and return the next one.

File: milabench/validation/nan.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+from .validation import ValidationLayer
+
+
+class _Layer(ValidationLayer):
+    """Makes sures the loss we receive is not Nan.
+
+    Notes
+    -----
+    Show a warning if the loss is not decreasing.
```
Comment (breuleux) [inline milabench/validation/nan.py:12]: Given that the validator's name is "nan", it's unexpected that it also checks that the loss does not increase. I think it would be better to rename the validator `loss`.

File: milabench/validation/nan.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            prev = self.previous_loss.get(tag)
+            self.previous_loss[tag] = loss
+
+            if prev is not None:
+                latest = int(math.isnan(loss))
+                self.warnings[tag]["nan_count"] += latest
+                self.nan_count += latest
+
+                if loss > prev:
+                    self.warnings[tag]["increasing_loss"] += 1
```
Comment (breuleux) [inline milabench/validation/nan.py:42]: I think it is also suspect if the loss does not change at all (possibly more than if it increases, at least at the beginning of training when it should certainly not be in a steady state). We could have a separate category for "loss remained the same").

File: milabench/validation/nan.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    def report(self, summary, **kwargs):
+        for bench, warnings in self.warnings.items():
```
Comment (breuleux) [inline milabench/validation/nan.py:45]: I believe nothing will be reported if no loss is reported at all. But that is also problematic.

File: milabench/validation/planning.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+        p.errors += int(p.method != method)
+        p.errors += int(p.njobs != njobs)
+        p.errors += int(method not in ("per_gpu", "njobs"))
+
+        # Counts the number of loss we are receiving
+        if entry.data:
+            loss = entry.data.get("loss")
+            if loss is not None:
+                p.loss[tag] = True
```
Comment (breuleux) [inline milabench/validation/planning.py:58]: Okay, I see that the presence of a loss is tracked here. It wouldn't be too much to also report it in the loss validator, though.

File: milabench/validation/usage.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+    """
+
+    def __init__(self, **kwargs) -> None:
+        self.warnings = defaultdict(lambda: defaultdict(float))
+
+        self.devices = set()
+        self.count = 0
+        self.mem_threshold = 0.50
+        self.load_threshold = 0.01
```
Comment (breuleux) [inline milabench/validation/usage.py:21]: We can leave it at that for now, but it'd be nice to make this configurable. I think the ideal place to do that would actually be in the benchmark's configuration, since we could adjust the expectations depending on the benchmark.

Also, `gpu_load/mem_threshold` would be a better naming convention. And we can track CPU/RAM as well in a next update.

File: milabench/validation/validation.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+from contextlib import contextmanager
+from dataclasses import dataclass, field
+
+from ..structs import BenchLogEntry
+
+
+class ValidationLayer:
+    """Validation layer interface, captures events, makes report"""
+
+    _rc = 0
```
Comment (breuleux) [inline milabench/validation/validation.py:11]: What does this do?

File: milabench/validation/validation.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+    def __init__(self, **kwargs) -> None:
+        self.early_stop = False
+        self._rc = 0
```
Comment (breuleux) [inline milabench/validation/validation.py:15]: I'm not sure that's a good default, I'd rather go with `None` to signify that it hasn't been set.

**Review by breuleux — COMMENTED** [review]
I haven't looked at tests yet.

File: milabench/validation/validation.py (hunk 11 above)
Comment (Delaunay) [reply milabench/validation/validation.py:11]: this is the return code of the validation layer, this is used to make milabench fail on critical errors..

File: milabench/validation/validation.py (hunk 12 above)
Comment (Delaunay) [reply milabench/validation/validation.py]: done

File: milabench/validation/usage.py (hunk 10 above)
Comment (Delaunay) [reply milabench/validation/usage.py]: done, added gpu_ to the thresholds

File: milabench/validation/planning.py (hunk 9 above)
Comment (Delaunay) [reply milabench/validation/planning.py:58]: done

File: milabench/validation/nan.py (hunk 8 above)
Comment (Delaunay) [reply milabench/validation/nan.py:46]: I added the warnings

File: milabench/utils.py (hunk 2 above)
Comment (Delaunay) [reply milabench/utils.py]: done, renamed to Layer and added the possibility to specify __layers__ as well

File: milabench/utils.py (hunk 1 above)
Comment (Delaunay) [reply milabench/utils.py]: done

File: milabench/utils.py [hunk 13]
```diff
@@ -102,3 +108,92 @@ def make_constraints_file(pth, constraints):
...
+        if layer is not None:
+            layers.append(layer())
+        else:
+            names = list(VALIDATION_LAYERS.keys())
+            raise RuntimeError(f"Layer `{layer_name}` does not exist: {names}")
+
+    return layers
+
+
+class _LoggerProxy:
```
[the remark this answers is not in the document]
Comment (Delaunay) [reply milabench/utils.py]: done

File: milabench/cli.py [hunk 14]
```diff
@@ -256,41 +255,55 @@ def _error_report(reports):
 
 
 def run_with_loggers(coro, loggers, mp=None):
+    # I feel we could clean up this code with ExitStack
+    # if loggers are contextmanagers
```
[the remark this answers is not in the document]
Comment (Delaunay) [reply milabench/cli.py]: yes removed

File: milabench/validation/ensure_rate.py (hunk 3 above)
Comment (Delaunay) [reply milabench/validation/ensure_rate.py:25]: It is necessary because if we are not receiving rates then when doing `for tag, rate in self.rates.items():` we will not be iterating through the missing benches that did not execute the `self.rates[tag] += entry.data.get("rate", 0)`.

I removed the defaultdict to prevent confusion in the future

File: milabench/validation/nan.py (hunk 6 above)
Comment (Delaunay) [reply milabench/validation/nan.py]: done

File: milabench/validation/validation.py (hunk 11 above)
Comment (breuleux) [reply milabench/validation/validation.py]: This is a class variable, though, but grep tells me it's only ever accessed through self.

File: milabench/validation/ensure_rate.py (hunk 3 above)
Comment (breuleux) [reply milabench/validation/ensure_rate.py:21]: Ah, I see.

File: milabench/validation/validation.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    def on_start(self, entry):
+        pass
+
+    def on_error(self, entry):
+        pass
+
+    def on_line(self, entry):
+        pass
+
+    def on_data(Self, entry):
```
Comment (breuleux) [inline milabench/validation/validation.py:49]:
```suggestion
    def on_data(self, entry):
```

File: milabench/validation/validation.py [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+        return self._return_code
+
+    def set_error_code(self, code):
+        self._return_code = code
+
+    def end(self):
+        return self._return_code
+
+    def on_event(self, entry: BenchLogEntry):
+        raise NotImplementedError(f"Event on_{entry.event} not implemented")
```
Comment (breuleux) [inline milabench/validation/validation.py:69]: Because all known events have default implementations that do nothing, this is only ever going to be called for new events that we add for some reason. I would rename `_on_event` to `on_event` and allow the validator to just override that if they want to do something generic (they can call super() to get the default behavior).

File: milabench/validation/validation.py [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+        self._return_code = None
+
+    def __call__(self, entry):
+        return self._on_event(entry)
+
+    def _on_event(self, entry: BenchLogEntry):
+        self.early_stop = False
+
+        if entry.event == "stop":
+            self.early_stop = True
```
Comment (breuleux) [inline milabench/validation/validation.py:29]: That could be the body of the default implementation of `on_stop`.

File: milabench/validation/validation.py [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+
+        # return code of the validation layer
+        # this is used to make milabench fail on critical errors
+        self._return_code = None
+
+    def __call__(self, entry):
+        return self._on_event(entry)
+
+    def _on_event(self, entry: BenchLogEntry):
+        self.early_stop = False
```
Comment (breuleux) [inline milabench/validation/validation.py:26]: Doesn't look like this should be here.

Comment (Delaunay) [reply milabench/validation/validation.py]: done

File: milabench/validation/validation.py (hunk 17 above)
Comment (Delaunay) [reply milabench/validation/validation.py]: done

File: milabench/validation/validation.py (hunk 16 above)
Comment (Delaunay) [reply milabench/validation/validation.py]: done

File: milabench/validation/ensure_rate.py (hunk 4 above)
Comment (Delaunay) [reply milabench/validation/ensure_rate.py]: done

File: milabench/validation/validation.py (hunk 11 above)
Comment (Delaunay) [reply milabench/validation/validation.py]: It was to make the value available through inheritance without having to call the super.__init__ to avoid changing too much code but I guess it is not needed right now.
I removed it

File: config/base.yaml [hunk 19]
```diff
@@ -5,6 +5,11 @@ _defaults:
       stop: 60
       interval: "1s"
 
+  validations:
```
[the remark this answers is not in the document]
Comment (Delaunay) [reply config/base.yaml]: done

[the remark this answers is not in the document]
Comment (Delaunay) [reply config/base.yaml]: done

File: milabench/validation/nan.py (hunk 7 above)
Comment (Delaunay) [reply milabench/validation/nan.py]: done

**Review by breuleux — APPROVED** [review]
